src/arithmetic/algebraic_expression.c

```diff
@@ -389,51 +390,66 @@ AlgebraicExpression **AlgebraicExpression_From_Query(const AST *ast, Vector *mat
     return expressions;
 }
 
+/* Evaluates an algebraic expression,
+ * evaluation is done right to left due to matrix CSC representation 
+ * the right most operand in the expression is a tiny extremely sparse matrix
+ * this allows us to avoid computing multiplications of large matrices.
+ * In the case an operand is marked for transpose, we will perform
+ * the transpose once and update the expression. */
 void AlgebraicExpression_Execute(AlgebraicExpression *ae, GrB_Matrix res) {
     assert(ae && res);
-
     size_t operand_count = ae->operand_count;
     assert(operand_count > 1);
 
-    AlgebraicExpressionOperand operands[operand_count];
-    memcpy(operands, ae->operands, sizeof(AlgebraicExpressionOperand) * operand_count);
-
-    GrB_Descriptor desc;
-    GrB_Descriptor_new(&desc);
     AlgebraicExpressionOperand leftTerm;
     AlgebraicExpressionOperand rightTerm;
+    // Operate on a clone of the expression.
+    AlgebraicExpressionOperand operands[operand_count];
+    memcpy(operands, ae->operands, sizeof(AlgebraicExpressionOperand) * operand_count);
 
-    /* Multiply right to left,
-     * A*B*C*D,
+    /* Multiply right to left
+     * A*B*C*D
      * X = C*D
      * Y = B*X
      * Z = A*Y */
     while(operand_count > 1) {
+        // Multiply and reduce.
         rightTerm = operands[operand_count-1];
         leftTerm = operands[operand_count-2];
 
-        // Multiply and reduce.
-        if(leftTerm.transpose) GrB_Descriptor_set(desc, GrB_INP0, GrB_TRAN);
-        if(rightTerm.transpose) GrB_Descriptor_set(desc, GrB_INP1, GrB_TRAN);
+        /* Incase we're required to transpose left hand side operand 
+         * perform transpose once and update original expression. */
+        if(leftTerm.transpose) {
+            GrB_Matrix t = leftTerm.operand;
+            /* Graph matrices are immutable, create a new matrix. 
+             * and transpose. */
+            if(!leftTerm.free) {
+                GrB_Index cols;
+                GrB_Matrix_ncols(&cols, leftTerm.operand);
+                GrB_Matrix_new(&t, GrB_BOOL, cols, cols);
```

Comment:
`t` is empty if this condition evaluates - I would expect the need for:
`GrB_Matrix_dup(&t, leftTerm.operand);`

Reply:
Indeed t is empty at this point,
it is being set at line 431, by GrB_transpose.